src/nitypes/scalar/_scalar.py

```diff
@@ -0,0 +1,123 @@
+from __future__ import annotations
+
+from typing import Generic
+
+from typing_extensions import final, TypeVar
+
+from nitypes._exceptions import invalid_arg_type, invalid_arg_value
+from nitypes.waveform._extended_properties import (
+    UNIT_DESCRIPTION,
+    ExtendedPropertyDictionary,
+)
+
+_ScalarType = TypeVar("_ScalarType", bool, int, float, str)
```

Comment:
Since the _value field is immutable, consider making the type variable covariant.